packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,191 @@
+
+const IntervalEventType = {
+	CHANGE: 0,
+	ADD: 1,
+	DELETE: 2,
+	PROPERTYCHANGED: 3,
+} as const;
+
+type IntervalEventType = typeof IntervalEventType[keyof typeof IntervalEventType];
+
+type IntervalRevertible =
+	| {
+			event: typeof IntervalEventType.CHANGE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.ADD;
+			interval: SequenceInterval;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.DELETE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.PROPERTYCHANGED;
+			interval: SequenceInterval;
+			propertyDeltas: PropertySet;
+	  };
+
+type TypedRevertible<T extends IntervalRevertible["event"]> = IntervalRevertible & { event: T };
+
+export function appendLocalAddToRevertibles(
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	revertibles.push({
+		event: IntervalEventType.ADD,
+		interval,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
+		end: interval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of previous interval
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: previousInterval.start,
+		end: previousInterval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	const label = revertible.interval.properties.label;
+	// should I use revertible.collection here or access from sharedstring?
```

Comment:
It's better not to have redundancy. If we don't need both, you can remove one.

Reply:
Not exactly sure what you mean by this comment. Is the redundancy having 2 different ways to access the intervalCollection, and if so, is it even possible to remove one of them?

Reply:
Yes. I don't think it's a big deal. It's possible to limit somewhat by not passing them into the functions that don't need them. We could reduce the chance of them being conflicting by storing the interval collection label instead of the collection itself. It seems hard to screw this up unless the collection is deleted in the meantime.

Reply:
In this specific case, you can remove the collection from the revertible.

Reply:
Should I do that for all of the revert methods then? Or should I keep the collection on revertible for add/delete/change and add it on the revertible passed to propertyChanged as well? If the cases are different enough, then I could remove it for this method and add it for propertyChanged, but I wasn't sure if they were.